smartsim/_core/_install/redisaiBuilder.py

```diff
@@ -0,0 +1,266 @@
+# BSD 2-Clause License
+#
+# Copyright (c) 2021-2024, Hewlett Packard Enterprise
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import fileinput
+import pathlib
+import re
+import shutil
+import subprocess
+import typing as t
+from collections import deque, namedtuple
+
+from packaging.version import Version
+
+from smartsim._core._cli.utils import SMART_LOGGER_FORMAT
+from smartsim._core._install.buildenv import BuildEnv
+from smartsim._core._install.mlpackages import MLPackageCollection
+from smartsim._core._install.platform import Platform
+from smartsim._core._install.utils import PackageRetriever
+from smartsim._core.config import CONFIG
+from smartsim.log import get_logger
+
+logger = get_logger("Smart", fmt=SMART_LOGGER_FORMAT)
+
+
+class RedisAIBuilder:
+    """Class to build RedisAI from Source"""
+
+    def __init__(
+        self,
+        platform: Platform,
+        mlpackages: MLPackageCollection,
+        build_env: BuildEnv,
+        verbose: bool = False,
+        source: t.Union[str, pathlib.Path] = "https://github.com/RedisAI/RedisAI.git",
+        version: str = "v1.2.7",
+    ) -> None:
+        self.platform = platform
+        self.mlpackages = mlpackages
+        self.build_env = build_env
+        self.verbose = verbose
+        self.source = source
+        self.version = version
+        self.patches: t.List[_RedisAIPatch] = []
+        self._define_patches_by_version()
+
+        self.src_path = CONFIG.build_path / "RedisAI" / "src"
+        self.build_path = CONFIG.build_path / "RedisAI" / "build"
+        self.package_path = CONFIG.build_path / "RedisAI" / "mlpackages"
```

Comment:
I'll use my discretion to decline this suggestion. One of the things that I would love to do in the future is freeze all these variables related to fundamental paths at pip install time and/or smart build time. It's difficult otherwise to ensure that the world hasn't changed underneath the install.